FreeBSD: Rework scope + building libc

rhelmot opened the pull request:
## Description of changes

**This PR depends on https://github.com/NixOS/nixpkgs/pull/298717**, however it is a little big so I would like to get started on reviews now.

This reworks the building of packages based on the FreeBSD source tree in order to support choosing a source branch. It adds the bare minimum number of packages to be able to cross compile the FreeBSD libc (`pkgsCross.x86_64-freebsd.freebsd.libc`), and it removes the major version number from the target tuple, as per prior discussion.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] x86_64-freebsd
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Mic92 commented:
Part of https://github.com/NixOS/nixpkgs/pull/296581

alyssais commented:
Please explain the reasons for changes in commit messages, not just the PR body, in as much detail as you can.  For core changes like this, it's really important to be able to look back and understand why things are the way they are, when there's a need to change them later.

> It adds the bare minimum number of packages to be able to cross compile the FreeBSD libc (`pkgsCross.x86_64-freebsd.freebsd.libc`)

I'm a bit confused by this, because this attribute already builds for me?

Ericson2314 commented:
The build just failed for me with 

```
/build/xinstall-bootstrap-filtered-src/lib/libnetbsd/sha2.h:38:10: fatal error: sha384.h: No such file or directory
```

I definitely like the `by-name` thing! I wonder if it makes sense to do that separately.

Ericson2314 commented:
> I definitely like the by-name thing! I wonder if it makes sense to do that separately, without any semantic changes.

I might actually take a stab at this, to help things along.

Ericson2314 commented:
https://github.com/NixOS/nixpkgs/pull/299347 OK an here it is. This should be a much simpler diff on top :).

Ericson2314 commented:
https://github.com/NixOS/nixpkgs/pull/299347 is now, merged, and I did a crude rebase on top

Ericson2314 commented:
Opened up #300707 for a next step

rhelmot commented on pkgs/os-specific/bsd/setup-hook.sh:
```diff
@@ -49,7 +49,6 @@ addMakeFlags() {
   makeFlags="LIBDIR=${!outputLib}/lib $makeFlags"
   makeFlags="SHLIBDIR=${!outputLib}/lib $makeFlags"
   makeFlags="SHAREDIR=${!outputLib}/share $makeFlags"
-  makeFlags="MANDIR=${!outputMan}/share/man $makeFlags"
```
hmmm - this seems to be an @artemist commit, https://github.com/rhelmot/nixpkgs/commit/3e2e56f462fc048e63f5d201f0b882536618adba. Files from it might be missing from this PR...

artemist replied:
FreeBSD treats it a bit differently. On NetBSD you might set `MANDIR=/usr/share/man` then put man files in `$(MANDIR)/man3`, on FreeBSD you'd set `MANDIR=/usr/share/man/man` then put man files in `$(MANDIR)3`
I added the original back to NetBSD but it may not have made it into this PR.

Ericson2314 replied:
OK makes sense

artemist committed:
freebsd: Cleanup, get ready to support version 14

* Extend libc

  Include non-libc core libraries in the libc package. Many of these
  mirror libraries present in glibc on linux, such as libgcc, libraries
  used for iconv, and libraries used for reading kernel info (libkvm,
  libprocstat, libmemstat).

  Without this many packages outside the freebsd tree would need to be
  modified to include standard dependencies which would already be on
  the system for other packages.

* Mark FreeBSD as using LLVM

* Update default LLVM version FreeBSD

* Use patch monolith

  The patchesRoot system combined with the fact that each derivation
  will Request specific names of patches makes it very annoying to use
  other FreeBSD source trees with nixpkgs. This new system allows
  providing one Or more entire trees of patches whose contents will be
  dynamically Parsed and only the relevant patches will be applied for
  any one Derivation.

  With this commit, the following knobs are available for specifying the
  FreeBSD source:

  - overriding `freebsd.versionInfo`, for picking another official
    supported FreeBSD release.

  - overriding `freebsd.source` for specifying a specific unpatched
    FreeBSD source tree.

  - overriding `freebsd.patches`, for specifying the patches to apply.

Co-Authored-by: Audrey Dutcher <audrey@rhelmot.io>
Co-Authored-by: John Ericson <John.Ericson@Obsidian.Systems>

artemist commented:
I rebased and added a better commit message. It would be rather challenging to split this more without breaking build on intermediate commits, so I'd rather not have to.

Mic92 reviewed (approved):
Looks good to me. @Ericson2314 any further concerns? Otherwise I would go ahead merging this.

Mic92 commented:
@ofborg eval

Ericson2314 commented on pkgs/os-specific/bsd/freebsd/patches/14.0/mtree-Makefile.patch:
```diff
@@ -0,0 +1,13 @@
+-DPADD+= ${LIBUTIL}
+-LDADD+= -lutil
++LIBADD+= ${LIBUTIL}
++#LIBADD+= -lutil
```
Do we want this line?

Ericson2314 commented:
There are some minor style things I would like to be different about this, I will try doing one myself, as it might be easier to demonstrate than explain.

Ericson2314 commented:
Also this didn't eval yet?

Ericson2314 commented:
`buildPackages.freebsdBranches.${hostBranch}` should be done with make scope with splicing instead of individual packages.

Perhaps we should set up video call or something to go over this in real life time? I feel like I'm just gonna do a really bad job conveying stuff async in this case, and so that will be easier.

Also, do we need zlib changes in this PR? That strikes me as unlikely.

rhelmot commented:
zlib changes moved to https://github.com/NixOS/nixpkgs/pull/307540.

Either I'm pretty fundamentally misunderstanding the things that you can do with splices, or I've done what you asked.

Ericson2314 commented:
The splicing is now set up properly. #308032 is the next step. I will probably pull out the `freebsdBranches` infra from this as step after that.

Ericson2314 commented:
OK I rebased this on top of the first few commits of #308138.

alyssais commented:
Last commit is still very difficult to review — the change to splice.nix could definitely be a separate commit, and if we're adding support for 14 while keeping support for 13, presumably that can be done one package at a time, or at least in smaller units than all at once?

Ericson2314 commented:
@alyssais I changed it to removing support for 13 based on your last comments, but we could do either. Also see #309577 which i made this depend on.

alyssais commented:
Ah, I see.  In that case I'll try to go through and review in the next few days, but any splitting that could still happen would make it easier, so I don't have to do it all in one go.

Ericson2314 commented:
Yeah I am trying to shrink the diff on this in general too (the formatting is just one part of that). I am marking it draft to make it clear that I haven't decided I am done with that yet.

Ericson2314 commented:
OK the diff has been minimized in a number of ways. The question to  @alyssais, @rhelmot and @artemist are:

1. Do we drop 13.0 and just do 14.0?
2. If so, can we get rid of all the `hostVersion` checks?

rhelmot commented:
I was actually working on simplifying this last night and then I woke up and you were messing with this branch! I would like to keep 13 support, but I can simplify it to just the major version as a differentiator and not any minor versions.

Ericson2314 commented:
@rhelmot sorry for the belated response! Yeah if you have use for 13, then sure that plan sounds good. Just wanted to be clear that I don't care about 13 / don't think because Nixpkgs was doing 13 Nixpkgs is insisting on supporting 13.

Ericson2314 commented:
I hope resurrecting your work simplifying after my rebasing and things will not be too hard :)

rhelmot commented:
I came up with one last simplification in order to improve the portability of the nixpkgs code across FreeBSD versions. The rationale and usage are in the most recent commit message.

Ericson2314 commented:
@rhelmot do you want to undo my deleting the version 13 patches? Also see the formatter failure --- I made formatted required for all things BSD recently.

rhelmot commented:
Yes, I didn't realize that was easily revertable. 

Could you explain how to run the formatter? I looked on the nixpkgs-fmt repo but couldn't find anything.

Ericson2314 commented:
(messaged in `#exotic:nixos.org` answering)

rhelmot commented:
There are no more changes I want to make before merge. Review time!

rhelmot committed:
freebsd: 13.1.0 -> 14.0.0

The old patches are kept because the version is actually configurable.
In the future both versions may be exposed.

Co-Auhtored-By: Artemis Tosini <me@artem.ist>
Co-Authored-by: John Ericson <John.Ericson@Obsidian.Systems>

Ericson2314 commented:
Looks great!! I just read redid the history but kept the final contents. So merging.
